fair/src/autogluon/fair/learners/fair.py

```diff
@@ -0,0 +1,418 @@
+                         metrics.per_group.median_absolute_error   (labels, y_pred, groups),
+                         metrics.per_group.spearmanr               (labels, y_pred, groups),
+                         metrics.per_group.pearsonr                (labels, y_pred, groups)))
+        scores=np.vstack ((base_metrics,r,scores))
+        gap=(scores.max(-1)-scores.min(-1)).reshape(-1,1)
+        names=('Expected label','Expected Prediction','r','r**2', 'RMS', 'mean AE',
+                 'median AE', 'Spearman r', 'pearson r'
+                 )
+        #print(collect,gap)
+        collect=np.hstack((scores,gap))
+        out=pd.DataFrame(collect.T,index =groups.unique().tolist()+ ["Maximum difference",],columns=names)
+        return out
+
+    def predict_proba(self,test_data):
+        "duplicates the functionality of classifier.predict_proba but with the new fair classifier"
+        proba=self.classifier.predict_proba(test_data)
+    
+        if isinstance(self.groups,str):
+            if self.infered_groups is False:
+                groups=np.asarray(test_data[self.groups])
+                
+        if self.infered_groups is False:
+            onehot=self.group_encoder.transform(groups.reshape(-1, 1)).toarray()
+        else:
+            onehot=np.asarray(self.infered_groups.predict_proba(test_data))
+        tmp=onehot.dot(self.offset)
+        proba+=tmp
+        return proba
+
+    def predict(self,test_data):
+        "duplicates the functionality of classifier.predict but with the new fair classifier"
+        proba=self.predict_proba(test_data)
+        return self.classifier.labels[proba.argmax(1)]
+
+
+def _needs_groups(fn):
+    "Internal helper function. Check if a metric requires a group argument or not"
+    #basically nothing works as some functions may take an implicit self or multiple keywords
+    #instead we just try to pass it two args and if it errors out return true
+    junk=np.zeros(1)
+    try:
+        fn(junk, junk)
+        return False
+    except Exception as e:
+        return True
+
+    #from inspect import getfullargspec
+    #sig=getfullargspec(fn)
+    #if sig[3] is None:
+    #    if len(sig[0])==2:
+    #        return False
+    #    return True
+    #return False
+
+def infered_attribute_builder(train,target,protected):
+    """Helper function that trains tabular predictors suitible for use when the protected attribute 
+        has to be infered, returns a classifier predicting the target that doesn't use the protected attribute
+        and a classifier predicting the protected attribute that doesn't use the target"""
+    target_train=train.drop(protected,axis=1,inplace=False)
+    protected_train=train.drop(target,axis=1,inplace=False)
+    target_classifier = autogluon.tabular.TabularPredictor(label=target).fit(train_data=target_train)
+    protected_classifier = autogluon.tabular.TabularPredictor(label=protected).fit(train_data=protected_train)
```

Comment:
Are we really going to just call TabularPredictor without any arguments to control for time_limit, models trained, etc? This won't be a good user experience.